### PR TITLE
check locally before pulling from registry

### PR DESCRIPTION
**Pull request by kevsmith**
This commit revamps Relay's image fetching logic to use a local image, if it exists, before pulling from the remote Docker registry. A special case for the tag `latest` is also introduced to ensure Relay always pulls in case image content has changed.

### CODE REVIEW

File: relay/engines/docker.go [hunk 1]
```diff
@@ -59,54 +59,55 @@ func (de *DockerEngine) Init() error {
...
-		ioutil.ReadAll(closer)
-		closer.Close()
-	}
-	if pullErr != nil {
-		log.Errorf("Error ocurred when pulling image %s: %s.", name, pullErr)
-		image, _, inspectErr := de.client.ImageInspectWithRaw(context.Background(), fullName, false)
-		if inspectErr != nil {
-			log.Errorf("Unable to find Docker image %s locally or in remote registry.", name)
-			return false, pullErr
+	found := false
```
Comment (christophermaier) [inline relay/engines/docker.go]: I don't think the `found` variable is ever touched after this... if so, at least one level of nesting can be removed from the whole function.

File: relay/engines/docker.go [hunk 2]
```diff
@@ -59,54 +59,55 @@ func (de *DockerEngine) Init() error {
...
+			_, removeErr := de.client.ImageRemove(context.Background(), beforeID,
+				types.ImageRemoveOptions{
+					Force:         true,
+					PruneChildren: true,
+				})
+			if removeErr != nil {
+				log.Errorf("Failed to remove old Docker image %s: %s.", shortImageID(beforeID), removeErr)
+			} else {
+				log.Infof("Replaced obsolete Docker image %s with %s.", shortImageID(beforeID), shortImageID(afterID))
+			}
```
Comment (christophermaier) [inline relay/engines/docker.go]: Might read easier to put a `return true, nil` here, particularly if we can remove the outer `if` as suggested above.

File: relay/engines/docker.go [hunk 3]
```diff
@@ -59,54 +59,55 @@ func (de *DockerEngine) Init() error {
...
 		} else {
-			log.Infof("Replaced obsolete Docker image %s with %s.", shortImageID(beforeID), shortImageID(afterID))
+			log.Infof("Docker image %s for %s is up to date.", shortImageID(beforeID), fullName)
```
Comment (christophermaier) [inline relay/engines/docker.go]: same here, re: return value

**Comment by vanstee** [comment]
Looks good.

If you have time it might be nice to break this down into a few function calls to make things more clear. It's not immediately obvious that `IsAvailable` would clean up old docker images for instance. Maybe renaming it `FetchImage` and then breaking it down into three functions like: `fetchLocalImage`, `fetchRemoteImage`, and `removeOldImages` would make this easier to read through.

**Pull request merged**